org/uacalc/alg/Malcev.java

```diff
@@ -732,6 +732,84 @@ public static Term fixedKEdgeTerm(SmallAlgebra alg, int k, ProgressReport report
     return null;
   }
 
+  /** 
+   * Alexandr Kazda's implementation (based on bits of previous UACalc code)
+   * of the local to global test for quasi weak near uninimity operation of
+   * arity <code>arity</code>. See the paper A. Kazda "Deciding the existence
+   * of quasi weak near unanimity terms in finite algebras"
+   *
+   * 
+   * @param testalg  an algebra 
+   * @param arity arity (at least 2)
+   * @return Boolean
+   */
+  public static boolean fixedKQwnu(SmallAlgebra testalg, int arity, ProgressReport report) {
+    if (arity < 2) throw new IllegalArgumentException("arity must be at least 2");
+    if (report != null) report.addStartLine("Using the local to global method to test for quasi WNU of arity " + arity + ". See the paper A. Kazda \"Deciding the existence of quasi weak near unanimity terms in finite algebras,\" 2020."); 
+    int power = arity;
+    BigProductAlgebra alg = new BigProductAlgebra(testalg, power);
+    final int size = testalg.cardinality();
+    //IntArray genvector = new IntArray(arity);
+    int a;
```

Comment:
Can these loop indices have more descriptive names? It looks like `a` is looping over the elements of `testalg`, and since `b` loops up to `a` I'm guessing it is too; but `i` sometimes loops over `testalg` and sometimes over `0..arity`.

Reply:
`i` now loops only over `0...arity` now. It specifies columns or positions on a diagonal of an `arity x arity` matrix.